pyresample/geometry.py

```diff
@@ -713,43 +715,301 @@ def copy(self):
+        # Check input validity
+        x = int(x)
+        y = int(y)
+        if x < 1 or y < 1:
+            raise ValueError("SwathDefinition.upsample expects (integer) upscaling factors >=1 .")
+        # Return SwathDefinition if nothing to upsample
+        if x == 1 and y == 1:
+            return self
+        # --------------------------------------------------------------------.
+        # TODO:
+        # - Refactor for dask-compatibility
+        # - Should we make _infer_interval_breaks dask-compatible?
+
+        def _get_corners_from_centroids(centroids):
+            breaks_xx = _infer_interval_breaks(centroids, axis=1)
+            corners = _infer_interval_breaks(breaks_xx, axis=0)
+            return corners
+
+        # TODO: choose one of the two function below
+        # - What is the best way to apply _upsample_centroid along each x-y plane with dask
+        def _upsample_centroid(centroid, x=1, y=1):
+            corners = _get_corners_from_centroids(centroid)
+            # Retrieve corners of the the upsampled grid
+            new_corners = _linspace2D_between_values(corners, num_x=x - 1, num_y=y - 1)
+            # Get centroids from corners
+            new_centroids = (new_corners[:-1, :-1] + new_corners[1:, 1:]) / 2
+            return new_centroids
+
+        def upsample_centroids(centroid_x, centroid_y, centroid_z, x=1, y=1):
+            x_new_centroids = _upsample_centroid(centroid_x, x=x, y=y)
+            y_new_centroids = _upsample_centroid(centroid_y, x=x, y=y)
+            z_new_centroids = _upsample_centroid(centroid_z, x=x, y=y)
+            return x_new_centroids, y_new_centroids, z_new_centroids
+
+        # --------------------------------------------------------------------.
+        # Define geodetic and geocentric projection
+        geocent = pyproj.Proj(proj='geocent')
+        latlong = pyproj.Proj(proj='latlong')
+
+        # Get xr.DataArray with dask array
+        # - If input lats/lons are xr.DataArray, the specified dims ['x','y'] are ignored
+        src_lons, src_lons_format = _convert_2D_array(self.lons, to='DataArray_Dask', dims=['y', 'x'])
+        src_lats, src_lats_format = _convert_2D_array(self.lats, to='DataArray_Dask', dims=['y', 'x'])
+
+        # Conversion to Geocentric Cartesian (x,y,z) CRS
+        res = da.map_blocks(self._do_transform, latlong, geocent,
+                            src_lons.data,
+                            src_lats.data,
+                            da.zeros_like(src_lons),  # altitude
+                            new_axis=[2],
+                            chunks=(src_lons.chunks[0], src_lons.chunks[1], 3))
+        res = xr.DataArray(res, dims=['y', 'x', 'xyz'])
+
+        # Retrieve new centroids
+        # TODO: make it dask compatible using _upsample_centroid_dask [HELP WANTED]
+        # res1 = da.apply_along_axis(_upsample_centroid_dask,
+        #                            2,
```

Comment:
@djhoese have a look at this portion of code when you will have half an hour.
I am not sure what is the best way to use dask here.

Reply:
I do not fully understand the logic here, but I understand the overall goal. To summarize we have an overall array of (M rows, N columns) and we want to produce a final array with a shape of (M * y, N * x), right? To do this we need convert from lon/lat to x/y/z geocentric, do some math, then convert back x/y/z to lon/lat. In the "do some math" portion, do we need to the values of all the pixels next to the current pixel? Or can we operate on groups of `y` by `x` pixels?

Given what I've learned about number of tasks in a dask graph, I think it would be best to do this entire thing in a giant map_blocks call. By "entire thing" I mean lon/lat -> x/y/z, the math, x/y/z -> lon/lat. If you require to work on groups of `y` by `x` pixels then you could make sure that the input chunks are divisible by those numbers (2x2 -> 1024x1024 would be ok). We don't want to make really small chunks. If you need overlap between the computations (in the case where each pixel needs to know about all the values around it) then `map_overlap` may be something to look into: https://docs.dask.org/en/stable/array-overlap.html